aten/src/ATen/core/type.cpp

```diff
@@ -760,6 +770,342 @@ TupleTypePtr TupleType::createNamed(
+  } else if (type->kind() == NumberType::Kind) {
+    types.emplace_back(IntType::get());
+    types.emplace_back(FloatType::get());
+    types.emplace_back(ComplexType::get());
+  } else {
+    types.emplace_back(type);
+  }
+}
+
+// Helper function for `standardizeUnion`. If we have types `T1`, `T2`,
+// and `T3` such that `T1` and `T2` are subtypes of `T3`, we want to
+// remove `T1` and `T2` from the `types` vector. Note that this is not
+// the same as `unifyTypes`, which will return the common supertype of
+// two types regardless of whether or not that supertype originally
+// existed
+void filterDuplicateSubtypes(std::vector<TypePtr>& types) {
+  auto get_supertype = [](const TypePtr t1, const TypePtr t2) -> c10::optional<TypePtr> {
+    if (t1->isSubtypeOf(t2)) {
+      return t2;
+    } else if (t2->isSubtypeOf(t1)) {
+      return t1;
+    } else {
+      return c10::nullopt;
+    }
+  };
+
+  // Coalesce types and delete all duplicates. Moving from right to left
+  // through the vector, we try to unify the current element (`i`) with
+  // each element (`j`) before the "new" end of the vector (`end`).
+  // If we're able to unify the types at `types[i]` and `types[j]`, we
+  // decrement `end`, swap `types[j]` with the unified type, and
+  // break. Otherwise, we keep `end` where it is to signify that the
+  // new end of the vector hasn't shifted
+  size_t end_idx = types.size()-1;
+  for (size_t i = types.size()-1; i > 0; --i) {
+    for (size_t j = std::min(i-1, end_idx); ; --j) {
+      c10::optional<TypePtr> unified;
+      unified = get_supertype(types[i], types[j]);
+      if (unified) {
+        types[j] = *unified;
+        types[i] = types[end_idx];
+        --end_idx;
+        break;
+      }
+      // Break condition here so we don't get `j = 0; j = j-1` and end
+      // up with MAX_INT
+      if (j == 0) {
+        break;
+      }
+    }
+  }
+  // Cut off the vector's tail so that `end` is the real last element
+  types.erase(types.begin() + end_idx + 1, types.end());
+}
+
+void standardizeUnion(std::vector<TypePtr>& types) {
+  // We want the elements to be sorted so we can easily compare two
+  // UnionType objects for equality in the future. Note that this order
+  // is guaranteed to be stable since we've already coalesced any
+  // possible types
+  std::sort(types.begin(), types.end(),
+          [](const TypePtr a, const TypePtr b) -> bool {
+            if (a->kind() != b->kind()) {
+              return a->kind() < b->kind();
+            }
+            return a->str() < b->str();
+          });
+}
+
+UnionType::UnionType(std::vector<TypePtr> types, TypeKind kind) : Type(kind) {
+  TORCH_INTERNAL_ASSERT(types.size() >= 2, "Cannot create a Union of "
+                        "one or fewer types");
+
+  for (auto type : types) {
+    flattenUnion(type, types_);
+  }
+
+  filterDuplicateSubtypes(types_);
+
+  standardizeUnion(types_);
+
+  TORCH_INTERNAL_ASSERT(types_.size() != 1, "After type unification was"
```

Comment:
I agree, but it's harder because this is the constructor (so, we inherently have to return a `UnionType`). I feel strongly that we should NOT allow "Unions" of a single type, since we would have to add in a ton of extra checks to ensure that the semantic remain the same